src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -3312,98 +3312,72 @@ class G1STWDrainQueueClosure: public VoidClosure {
+      _serial_keep_alive(&g1h, _pss.state_for_worker(0)),
+      _serial_complete_gc(&g1h, _pss.state_for_worker(0)),
+      _tm(ThreadModel::Single) {}
+
+  ~G1STWRefProcClosureContext() {
+    FREE_C_HEAP_ARRAY(G1CopyingKeepAliveClosure, _keep_alive);
+    FREE_C_HEAP_ARRAY(G1ParEvacuateFollowersClosure, _parallel_complete_gc);
+  }
+
+  BoolObjectClosure* is_alive(uint worker_id) { return &_is_alive; }
+  OopClosure* keep_alive(uint worker_id) {
+    assert(worker_id < _queues || _tm == ThreadModel::Single, "sanity");
+    if (_tm == ThreadModel::Single) {
+      return ::new (&_serial_keep_alive) G1CopyingKeepAliveClosure(&_g1h, _pss.state_for_worker(0));
+    }
+    return ::new (&_keep_alive[worker_id]) G1CopyingKeepAliveClosure(&_g1h, _pss.state_for_worker(worker_id));
   }
-
-  // Executes the given task using concurrent marking worker threads.
-  virtual void execute(ProcessTask& task, uint ergo_workers);
-};
-
-// Gang task for possibly parallel reference processing
-
-class G1STWRefProcTaskProxy: public AbstractGangTask {
-  typedef AbstractRefProcTaskExecutor::ProcessTask ProcessTask;
-  ProcessTask&     _proc_task;
-  G1CollectedHeap* _g1h;
-  G1ParScanThreadStateSet* _pss;
-  G1ScannerTasksQueueSet* _task_queues;
-  TaskTerminator* _terminator;
-
-public:
-  G1STWRefProcTaskProxy(ProcessTask& proc_task,
-                        G1CollectedHeap* g1h,
-                        G1ParScanThreadStateSet* per_thread_states,
-                        G1ScannerTasksQueueSet *task_queues,
-                        TaskTerminator* terminator) :
-    AbstractGangTask("Process reference objects in parallel"),
-    _proc_task(proc_task),
-    _g1h(g1h),
-    _pss(per_thread_states),
-    _task_queues(task_queues),
-    _terminator(terminator)
-  {}
-
-  virtual void work(uint worker_id) {
-    // The reference processing task executed by a single worker.
-    ResourceMark rm;
-
-    G1STWIsAliveClosure is_alive(_g1h);
-
-    G1ParScanThreadState* pss = _pss->state_for_worker(worker_id);
-    pss->set_ref_discoverer(NULL);
-
-    // Keep alive closure.
-    G1CopyingKeepAliveClosure keep_alive(_g1h, pss);
-
-    // Complete GC closure
-    G1ParEvacuateFollowersClosure drain_queue(_g1h, pss, _task_queues, _terminator, G1GCPhaseTimes::ObjCopy);
-
-    // Call the reference processing task's work routine.
-    _proc_task.work(worker_id, is_alive, keep_alive, drain_queue);
-
-    // Note we cannot assert that the refs array is empty here as not all
-    // of the processing tasks (specifically phase2 - pp2_work) execute
-    // the complete_gc closure (which ordinarily would drain the queue) so
-    // the queue may not be empty.
+  VoidClosure* complete_gc(uint worker_id) {
+    assert(worker_id < _queues || _tm == ThreadModel::Single, "sanity");
+    if (_tm == ThreadModel::Single) {
+      return ::new (&_serial_complete_gc) G1STWDrainQueueClosure(&_g1h, _pss.state_for_worker(0));
```

Comment:
Not sure why this closure needs to be reinitialized to the same values as in the constructor. Just return `&_serial_complete_gc` should be fine. I understand that for the parallel case, as they haven't been initialized yet.

Reply:
I will remove the placement new.